ch4: active message shared memory

wesbland opened the pull request:
This is a rebasing of #2526 along with a bunch of bug fixes. I'm going to preserve the original PR since there have been so many changes and start a new one here.

This branch is based on the branch in #3195. Not all of the patches are original in this PR.

Obviously, this PR is nowhere near ready to merge, but I'm putting it up here so we can work on it together. Here's a checking of things I know need to be changed or checked:

* [x] Fix non-direct builds
* [x] Wildcard receives need to be validated
* [x] The later patches are mostly "fixup" patches that should be squashed with the first patch. Some of the middle patches probably make sense to be on their own.
* [x] The big first patch might need to be split into smaller patches to avoid having such a monolithic big patch.
* [x] Make sure the inlining is correct
* [x] The last time I ran the tests, `pt2pt/sendrecv1 4` was segfaulting in auto mode
* [x] We need to decide what to do with the `__asm__ __volatile__` in `fbox_send.h`

CC: @raffenet @ddurnov @SergeyGubanov @minsii

wesbland commented:
test:jenkins/ch4/ofi

wesbland commented:
test:intel:ch4/ofi/sockets,psm2/all/all/all/debug,default

wesbland commented:
test:intel:ch4/ofi/sockets/gnu/all/auto/debug,default

wesbland commented:
test:intel:ch4/ofi/sockets/gnu/all/auto/debug,default

wesbland commented:
test:intel:ch4/ofi/sockets/gnu/all/auto/debug,default

wesbland commented:
test:intel:ch4/ofi/sockets/gnu/all/auto/debug,default

wesbland commented:
test:intel:ch4/ofi/sockets/gnu/all/auto/debug,default

wesbland commented:
test:intel:ch4/ofi/sockets/gnu/all/auto/debug,default

wesbland commented:
I reorganized the patches and squashed them down into two commits. One that covers the fixes for `MPI_ANY_SOURCE` that are outside of the shared memory code and one that covers everything inside the shared memory code.

It might be possible to split up the second patch, but I'm no longer convinced of that and the amount of work goes beyond what I care about.

I still have the [old branch](https://github.com/wesbland/mpich/tree/shm_am_not-reordered) around if it appears that something went wrong with my rebasing.

wesbland commented:
test:ch4/ofi/sockets/gcc/all/all/debug,default

wesbland commented:
test:intel:ch4/ofi/sockets/gnu/all/all/debug,default

wesbland commented:
At this point, I think this branch should be based on #3027. I'm going to rebase it onto that and remove the patch to ignore the RMA tests. It also needs #3235 so I'll cherry-pick that in.

wesbland commented:
Only the last two patches are specifically from this PR.

wesbland commented:
test:intel:ch4/ofi/sockets/gnu/all/all/debug,default

wesbland commented:
test:intel:ch4/ofi/sockets/gnu/all/all/debug,default

wesbland commented:
test:intel:ch4/ofi/sockets/gnu/all/all/debug,default

wesbland commented:
test:intel:ch4/ofi/sockets/gnu/all/all/debug,default

arekay commented:
test:intel:ch4/ofi/sockets/gnu/all/all/debug,default

wesbland commented:
test:intel:ch4/ofi/sockets/gnu/all/all/debug,default

wesbland commented:
test:intel:ch4/ofi/sockets/gnu/all/all/debug,default

wesbland commented:
test:intel:ch4/ofi/sockets/gnu/all/all/debug,default

wesbland commented:
test:intel:ch4/ofi/sockets/gnu/all/all/debug,default

wesbland commented:
test:intel:ch4/ofi/sockets/gnu/all/all/debug,default

wesbland commented:
test:intel:ch4/ofi/sockets/gnu/all/all/debug,default

wesbland commented:
test:intel:ch4/ofi/sockets/gnu/all/all/debug,default

minsii commented:
@wesbland : Could you please cherry pick the following commits from #3261 : 
- ch4: directly pass locality info to target_msg_cbs
- ch4/rma: store is_local info in callback-created request
- ch4/rma: reuse is_local in callback routines
- ch4/rma: return correct iov at target put callback
- ch4/rma: remove unnecessary copy in put AM callback

They should fix some RMA bugs.

Besides, I found that some hanging problems in RMA are not caused by large message, but caused by `MPIDI_POSIX_am_send_hdr`. It does not enqueue uncompleted  send request (i.e., no available cell in iqueue, or fbox is being used). This function is used only in RMA active message.
```
    result = MPIDI_POSIX_eager_send(grank, &msg_hdr_p, &iov_left_ptr, &iov_num_left, 1);
    mpi_errno = (result == MPIDI_POSIX_OK) ? MPI_SUCCESS : MPI_ERR_OTHER;
```

As a temporary workaround, it should be OK if we internally direct to `MPIDI_NM_am_send_hdr` or use `MPIDI_SHM_am_isend` instead. Otherwise, we need add the enqueue logic in `MPIDI_POSIX_am_send_hdr`. Please let me know your thought.

wesbland commented:
I've cherry-picked those patches over.

wesbland commented:
> As a temporary workaround, it should be OK if we internally direct to MPIDI_NM_am_send_hdr or use MPIDI_SHM_am_isend instead. Otherwise, we need add the enqueue logic in MPIDI_POSIX_am_send_hdr. Please let me know your thought.

I think adding the enqueue logic to `MPIDI_POSIX_am_send_hdr` is probably the right way to go. It's probably a little more work, but we're waiting for rendezvous anyway so let's do it right.

wesbland commented:
Per the conversation on https://github.com/pmodels/mpich/pull/3261#issuecomment-411151092, we're trying to decide whether to make `fbox` or `iqueue` the default `posix` submodule. @minsii is concerned that `fbox` will actually hurt message rate because it blocks as soon as the fast box is full.

I'm fine with making `iqueue` default (I don't have a strong opinion either way). What do you think @ddurnov / @SergeyGubanov?

minsii commented:
@wesbland I just did a force push:
- Fixed a previous commit "ch4/rma: remove unnecessary copy in put AM callback" for RMA bug
- Removed commit "ch4/rma: warning fix (temp)", which is already solved in master
- Temporarily enable `iqueue` in commit "temp: Make iqueue default posix/eager module"

minsii commented:
test:jenkins/ch4/ofi

minsii commented:
test:jenkins/ch4/ucx

minsii commented:
test:jenkins/ch4/ofi

minsii commented:
test:jenkins/ch4/ucx

wesbland commented:
@minsii - I'd like to drop:

    temp: use NM_am_send_hdr in ch4/shm

In favor of the solution we talked about previously where we actually enqueue headers properly. Is that ok? Do you know what needs to be implemented there?

minsii commented:
@wesbland : We need implement the queueing logic for the `send_hdr*` functions. I added the temp commit here just to make sure if any other failure exists and is not caused by this problem. Is anyone in your team working on the queueing logic now ? If not, I can work on it later today.

minsii commented:
test:jenkins/ch4/most

minsii commented:
Force pushed again to replace `temp: use NM_am_send_hdr in ch4/shm` with real queuing logic in `am_send_hdr`.

minsii commented:
test:jenkins/ch4/ofi

minsii commented:
@wesbland The queuing fix seems correctly handled RMA failures. Do you think anything else we need to work on except the rendezvous impl ?

wesbland commented:
test:intel:ch4/ofi/sockets/gnu/all/all/debug,default

wesbland commented:
test:intel:ch4/ofi/sockets/gnu/all/all/debug,default

wesbland commented:
test:intel:ch4/ofi/sockets/gnu/all/all/debug,default

wesbland commented:
I did a quick run of `gather_big` with the bandaid fix locally and everything passed after @minsii added her patch to correctly queue header messages. If the larger test suite passes, I might be more inclined to move forward on this without the full rendezvous implementation as things wouldn't be worse than they are right now.

wesbland commented on src/mpid/ch4/shm/posix/posix_init.h:
```diff
@@ -105,11 +107,14 @@ static inline int MPIDI_POSIX_mpi_init_hook(int rank, int size, int *n_vnis_prov
     }
 
+    /* XXX - Is the eager submodule never inlined? We might want to do that for certain situations,
+     * for instance, if only one eager submodule is compiled in. */
```
Is the eager submodule never inlined? We might want to do that for certain situations. For instance, if only one eager submodule is compiled in.

wesbland commented on src/mpid/ch4/shm/posix/posix_pre.h:
```diff
@@ -106,6 +106,8 @@ typedef struct MPIDI_POSIX_am_request_t {
 
     MPIDI_POSIX_am_request_header_t *req_hdr;
+
+/* XXX - Why is this ifdef here? I don't see it ever being defined. */
```
Why is this `ifdef` here? I don't see it ever being defined.

wesbland commented on src/mpid/ch4/shm/posix/posix_progress.h:
```diff
@@ -42,12 +42,15 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_POSIX_progress(int blocking)
     }
 
+    /* Process the eager message */
+    /* XXX - Should this not happen in an eager-specific progress engine? */
```
Should this not happen in an eager-specific progress engine?

wesbland commented:
BTW for those following, this is the current status of the tests:

![screen shot 2018-08-28 at 9 06 41 am](https://user-images.githubusercontent.com/1561207/44728317-c49ed400-aaa1-11e8-92ec-535473ba2607.png)

We're pretty confident that the problem here is _not_ the lack of a more optimized rendezvous protocol (otherwise other large tests would fail), but some other bug. Still working on it.

wesbland commented:
With the latest patches, all of the above tests are now passing. This is ready for a review from @raffenet and @minsii.

There are still two failures when forcing AM mode, but those are related to the same issues that @ddurnov is having with implementing rendezvous where the AM handlers are being set by NM and SHM code at the same time. When we resolve that issue, it should solve both problems.

I propose that this isn't a use case we're currently focused on (though it might become more important with the ideas in #2985) and that we can move forward on this in the meantime.

SergeyGubanov commented on src/mpid/ch4/shm/posix/eager/iqueue/iqueue_init.h:
```diff
@@ -46,6 +46,9 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_POSIX_eager_init(int rank, int size)
 
+    /* Get the internal data structure to describe the iqueues */
+    /* XXX - Why is this not just a global struct? Why is it abstracted inside this call? Do we ever
+     * need more than one of these? */
```
This abstraction layer was prepared for ability to allocate the transport structure by demand if it is needed and where it is needed (for example, on a huge page). It is possible to have many number (50-100) of different shm transport modules in one binary but only one of them actually used in run time (or shm may be disabled in run time at all). I believe, to many static allocated unused variables are not desired. 

Also, such abstraction layer may be helpful for future multiple threading shm transport implementation (it is not designed yet), in this scenario the "transport" pointer will be a thread local variable: each thread inside rank may have its own "transport" structure.

I agree that in the current implementation this additional abstraction layer may looks bizarre.

wesbland replied:
Ok. I guess I don't have a strong preference here then.

SergeyGubanov commented on src/mpid/ch4/shm/posix/posix_am_impl.h:
```diff
@@ -48,12 +48,15 @@ static inline int MPIDI_POSIX_am_init_req_hdr(const void *am_hdr,
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_POSIX_AM_INIT_REQ_HDR);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_POSIX_AM_INIT_REQ_HDR);
 
+    /* XXX - I don't know why this inline is here. */
```
The POSIX_AM_REQUEST_INLINE symbol turns on a feature which allowed lightweight (fast) allocation of the "req_hdr" object instead of expensive allocation one via MPIDI_CH4R_get_buf.

Usage of MPIDI_CH4R_get_buf has bad effect on message rate performance.

wesbland replied:
But this only seems to be set if you do it via `-D` at compile time, correct?

wesbland replied:
Would it make sense to make this a configure flag?

SergeyGubanov replied:
Yes, it is correct. If this feature considered helpful, it can be turned on by adding "-D POSIX_AM_REQUEST_INLINE" in the CFLAGS list.

wesbland commented:
I've rebased the PR, renamed the WIP patch and taken out the comments that I'd moved over to GitHub. This should be more ready for a review from Argonne, though I think we probably need to decide whether to squash it down to a single patch or not as we might still want to do some massaging of the commits if we don't want to squash them.

wesbland commented:
test:jenkins/ch4/most

kkandalla commented:
Hi Wes, Thanks for commenting on some of the performance trends with the new impl. If I understood you correctly during last week's call -- when compared to the current implementation in CH4/shm/posix,  you see up to 2x improvement for small message latency (<16KB). But,  large msg latency is worse. Can you also share the full comparisons for the  latency, bw, and bibw tests?

wesbland commented:
Sorry @kkandalla. I haven't had a chance to run all of those numbers yet. I got pulled off on other things.

I can try to have something for the call this week that's a little more fleshed out though.

minsii commented:
@wesbland The current PR seems too large. Do you think we can separate some commits into small PRs so we can review & merge one by one ?

wesbland commented:
@minsii I'm not sure that we can separate it. I wanted to do that when I started working on this PR as well, but I couldn't figure out a way to do a clean break. The problem is that the part that hooks into the AM code requires the part that refactors the shared memory code. The only thing that could feasibly be split into another patch is the iqueue code, but I'm not sure that's really all that beneficial.

The thing I was considering most recently is whether to just squash this whole thing down into one patch (or maybe two patches where `ch4: Fix any_source receives` is its own patch). Some of the later patches are sort of fixups of the first big `POSIX shared memory new logic` patch.

minsii commented:
The latest jenkins result shows some failures in ch4/ofi. Are you fixing them now ?

wesbland commented:
I wasn't. I'll go look at them.

wesbland commented:
Oh...those were because of the lack of rendezvous as mentioned above.

> There are still two failures when forcing AM mode, but those are related to the same issues that @ddurnov is having with implementing rendezvous where the AM handlers are being set by NM and SHM code at the same time. When we resolve that issue, it should solve both problems.
> 
> I propose that this isn't a use case we're currently focused on (though it might become more important with the ideas in #2985) and that we can move forward on this in the meantime.

These should be xfailed.

wesbland commented:
I rebased this onto the latest `master` so the test results disappeared. I think they'd been purged on Jenkins anyway so I'll run them again.

test:jenkins/ch4/most

wesbland commented:
So almost all of the ch4-ofi failures are expected because of the am handler problem. I'm not 100% sure about the `anysource` tests. I'll have to look at that again.

I don't know what to tell you about UCX. I have a hard time believing those tests since they get false positives a lot.

minsii commented on src/mpid/ch4/generic/mpidig_send.h:
```diff
@@ -70,7 +78,11 @@ static inline int MPIDI_am_isend(const void *buf, MPI_Aint count, MPI_Datatype d
+         * back from the receiver to indicate completion. */
         MPIR_cc_incr(sreq->cc_ptr, &c);
+
         if (MPIDI_av_is_local(addr)) {
```
Looks like a white space change.

minsii replied:
Whitespace still here ?

minsii commented on src/mpid/ch4/shm/posix/Makefile.mk:
```diff
@@ -5,34 +5,39 @@
 ##  Portions of this code were written by Intel Corporation.
-##  Copyright (C) 2011-2016 Intel Corporation.  Intel provides this material
+##  Copyright (C) 2011-2017 Intel Corporation.  Intel provides this material
 ##  to Argonne National Laboratory subject to Software Grant and Corporate
```
Should the copyright change be moved to a separate commit ?

minsii commented on src/mpid/ch4/shm/posix/Makefile.mk:
```diff
@@ -5,34 +5,39 @@
-                  src/mpid/ch4/shm/posix/posix_startall.h  \
-                  src/mpid/ch4/shm/posix/posix_unimpl.h
+                  src/mpid/ch4/shm/posix/posix_startall.h
 
```
Looks like some lines are reordered. Is it necessary ?

wesbland replied:
Ok.

minsii commented on src/mpid/ch4/shm/posix/posix_impl.h:
```diff
@@ -4,7 +4,7 @@
  *  Portions of this code were written by Intel Corporation.
- *  Copyright (C) 2011-2016 Intel Corporation.  Intel provides this material
+ *  Copyright (C) 2011-2017 Intel Corporation.  Intel provides this material
  *  to Argonne National Laboratory subject to Software Grant and Corporate
```
Should we put copyright change in this commit ?

wesbland replied:
I think it makes sense to update them at the same time.

minsii commented on src/mpid/ch4/generic/mpidig_recv.h:
```diff
@@ -54,6 +54,8 @@ static inline int MPIDI_handle_unexpected(void *buf,
 
+    /* Calculate the size of the data from the active message information and update the request
+     * object. */
     in_data_sz = MPIDI_CH4U_REQUEST(rreq, count);
```
Should we put ch4 comment enhancement in this commit ?

wesbland replied:
This could potentially move over to the prereq branch. I had a patch that added a bunch of comments, but it also included some non-posix stuff.

wesbland replied:
I've moved this change to the prereq branch. It still shows up here, but as a separate patch.

minsii commented:
Can the reordering change be moved to a different commit ?

Min

Sent from my iPhone

On Sep 19, 2018, at 1:22 PM, Wesley Bland <notifications@github.com<mailto:notifications@github.com>> wrote:


@wesbland commented on this pull request.

________________________________

In src/mpid/ch4/shm/posix/Makefile.mk<https://github.com/pmodels/mpich/pull/3199#discussion_r218912570>:

>                    src/mpid/ch4/shm/posix/posix_win.h       \
-                  src/mpid/ch4/shm/posix/posix_comm.h      \
-                  src/mpid/ch4/shm/posix/posix_defs.h      \
-                  src/mpid/ch4/shm/posix/posix_impl.h      \
-                  src/mpid/ch4/shm/posix/posix_probe.h     \
-                  src/mpid/ch4/shm/posix/posix_queue.h     \
-                  src/mpid/ch4/shm/posix/posix_request.h   \
-                  src/mpid/ch4/shm/posix/posix_send.h      \
-                  src/mpid/ch4/shm/posix/posix_startall.h  \
-                  src/mpid/ch4/shm/posix/posix_unimpl.h
+                  src/mpid/ch4/shm/posix/posix_startall.h



This keeps them (almost) in alphabetical order. I think that's useful, though posix_startall.h needs to move up to the right place.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/pmodels/mpich/pull/3199#discussion_r218912570>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AA272uOE5alSTpbZtvngLm_uYX38WbVCks5ucotkgaJpZM4U21Im>.

wesbland commented:
I don't think so. We already have to add a bunch of files because of the new code in the posix submodule. Separating that to two commits is weird and artificial.

wesbland commented:
test:jenkins/ch4/most

minsii commented on src/mpid/ch4/src/ch4r_callbacks.h:
```diff
@@ -456,23 +456,22 @@ static inline int MPIDI_send_target_msg_cb(int handler_id, void *am_hdr,
-        while (TRUE) {
+        int continue_matching = 1;
+        while (continue_matching) {
             rreq = MPIDI_CH4U_dequeue_posted(hdr->src_rank, hdr->tag, hdr->context_id,
```
Can this change also be applied to other wildcard_recv handling code in the callback file ? Actually, why not just squash them with the previous commit ?

wesbland commented on src/mpid/ch4/shm/posix/posix_init.h:
```diff
@@ -4,231 +4,119 @@
-        MPIR_Assert(MPIDI_POSIX_ALIGNED
-                    (MPIDI_POSIX_mem_region.RecvQ[grank], MPIDI_POSIX_CACHE_LINE_LEN));
+    for (i = 0; i < size; i++) {
+        MPIDI_POSIX_global.active_rreq[i] = NULL;
```
I'm not sure what you mean. Aren't _all_ processes shm processes?

wesbland replied:
I'm confused. I don't think that's what `active_rreq` does. This is an O(p) array of requests that need to be progressed.

minsii replied:
Yes, it is used in progress. But it's indexed by global rank. The reason I guess it uses global rank instead of shm_rank is to saving translation overhead.

wesbland replied:
Ok. I see what you mean.

In that case, I support the idea that we add this optimization later.

wesbland commented on src/mpid/ch4/shm/posix/posix_init.h:
```diff
@@ -333,60 +209,4 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_POSIX_create_intercomm_from_lpids(MPIR_Comm *
-    return mpi_errno;
-}
-
 #endif /* POSIX_INIT_H_INCLUDED */
```
Yeah, I think that makes sense here.

wesbland commented on src/mpid/ch4/shm/posix/posix_pre.h:
```diff
@@ -32,17 +104,78 @@ typedef struct {
+                             "**windows_mutex %s", "MPL_proc_mutex_unlock");        \
+} while (0)
+
 #include "posix_coll_params.h"
```
Yep. You're right. I'll split that out.

wesbland commented on src/mpid/ch4/generic/mpidig_recv.h:
```diff
@@ -128,7 +128,7 @@ static inline int MPIDI_do_irecv(void *buf,
 
-    root_comm = MPIDI_CH4U_context_id_to_comm(comm->recvcontext_id);
+    root_comm = MPIDI_CH4U_context_id_to_comm(context_id);
     unexp_req = MPIDI_CH4U_dequeue_unexp(rank, tag, context_id,
```
I think this was intended to just fix a problem with `MPI_ANY_SOURCE`. I don't know if it had any other side effects (we didn't see them in the testing).

hajimefu replied:
There has been many questions about this change. @SergeyGubanov it would be very helpful if you could explain this change.

wesbland commented on src/mpid/ch4/src/ch4_recv.h:
```diff
@@ -57,14 +57,19 @@ MPL_STATIC_INLINE_PREFIX int MPID_Recv(void *buf,
+            MPIDI_CH4I_REQUEST(MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(*request), is_local) = 0;
+
+            /* cancel the shm request if netmod/am handles the request from unexpected queue. */
             if (MPIR_Request_is_complete(MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(*request))) {
```
There are in progress requests in the active message code (this fix was not just for shared memory but for all active message code). Look in `mpidig_recv.h`.

wesbland replied:
There isn't anything in the netmod code (or the shared memory code for that matter) independently. It's all in the active message code.

wesbland replied:
It doesn't work the other way around because we call the shared memory code first every time (because it's software matching so cancellation is easier).

wesbland commented on src/mpid/ch4/src/ch4r_callbacks.h:
```diff
@@ -114,14 +118,61 @@ static inline int MPIDI_handle_unexp_cmpl(MPIR_Request * rreq)
+                    goto fn_fail;
+                }
+                /* What should we do if the anysource partner request is not canceled? */
+                MPIR_Assertp(MPIR_STATUS_GET_CANCEL_BIT(anysource_partner->status));
```
Not sure. This is an odd comment. I think it should be removed.

wesbland replied:
I'm going through and adding a bunch of comments for this part of the code, but the gist of this part is this:

Unlike the case below, this should always be true because the request has already been matched. If the partner request can't be cancelled, then the request has been double-matched and we have a problem.

minsii commented on src/mpid/ch4/shm/posix/posix_pre.h:
```diff
@@ -32,17 +104,78 @@ typedef struct {
+        MPIDI_POSIX_eager_recv_completed_hook((request)->dev.ch4.shm.posix.eager_recv_posted_hook_grank); \
+    } \
+    (request)->dev.ch4.shm.posix.eager_recv_posted_hook_state = MPIDI_POSIX_EAGER_RECV_POSTED_HOOK_STATE_FINALIZED; \
+} while (0)
```
I am not sure why we need these HOOK macros. Are they just used for state tracking ?  Can we disable it when `--enable-fast` is set ?

SergeyGubanov replied:
It is a key feature. It is optimization to minimize number of processed unexpected messages. But this key feature doesn't used in simple shm transports like iqueue and fbox.

An advanced shm transport (for example shm transport used in the Intel MPI 2019) must have information about expected or unexpected an available message (request already posted or not yet). If the request was not being posted yet, an advanced shm transport may start process another (expected) message from another rank. Messages from different ranks can be reordered, processing of unexpected message can be retarded.

minsii replied:
If it is not used in current `fbox` or `iqueue`, we should exclude it in this PR. When we add future transport that requires this feature , we can add the code back.

wesbland replied:
I tend to agree with @minsii here. Until it's needed in the public version of MPICH, it probably shouldn't be here.

wesbland replied:
It does look like the hook functions are necessary however. I guess we just don't need these macros.

wesbland replied:
So it looks like the `MPIDI_POSIX_EAGER_RECV_POSTED_HOOK` (and the function it calls, `MPIDI_POSIX_eager_recv_posted_hook`) _is_ actually used in the fbox transport. I'm not sure we should remove this code after all.

minsii replied:
It looks like `next_poll_local_rank` is being used in the fbox transport. Can we only keep this part of code in the PR ? The status changes seem unnecessary.

wesbland replied:
I think the state change is necessary in order to make sure that doing the other things are valid. The state change is only used inside these macros, but it still acts as a useful check.

wesbland replied:
I think you're right. The state change might not be necessary. It would be nice to keep the macros though because of the translation from communicator rank to global rank.

I think the state should be removed (and removed from the request structure) and the rest should stay.

minsii commented on src/mpid/ch4/shm/posix/eager/iqueue/iqueue_types.h:
```diff
@@ -0,0 +1,97 @@
+    uint32_t payload_size;      /* Size of the message in the cell */
+    uintptr_t prev;             /* Internal pointer when finding the previous active cell */
+    MPIDI_POSIX_EAGER_IQUEUE_cell_t *next;      /* Internal pointer when finding the next active cell */
+    MPIDI_POSIX_am_header_t am_header;  /* If this cell is the beginning of a message, it will have
```
Why `prev` is `uintptr_t ` but `next` is `MPIDI_POSIX_EAGER_IQUEUE_cell_t ` ?

wesbland replied:
Prev is an index in the array while next is a pointer to the cell itself. I don't know why that is right now. I'll have to go try to figure it out again.

SergeyGubanov replied:
The "prev" is the uintptr_t because it is the absolute offset of a cell in the shared memory segment. Offset is absolute, the same over all ranks. The field "prev" architecturally has a public scope of visibility shared between all ranks.

The "next" is a usual pointer because it is a private process memory virtual address. Different process have different virtual addresses of the same shared memory cell. The field "next" architecturally has a private scope of visibility inside the rank which in the current moment process the cell.

minsii replied:
I see. Could you please add short description in code comment ?

wesbland replied:
Done.

minsii commented on src/mpid/ch4/shm/posix/eager/iqueue/iqueue_send.h:
```diff
@@ -0,0 +1,136 @@
+                  /* Swaps the head of the terminal with the current handle if the previous head has
+                   * now been consumed. Continues until we swap out the prev pointer. */
+                  OPA_cas_ptr((OPA_ptr_t *) & terminal->head, (void *) prev, (void *) handle)
+                  != prev));
```
Moreover, why the `cell->next` pointer  is not set  ?

wesbland replied:
> Casting `uintptr_t *` to `OPA_ptr_t` seems unsafe. We don't know how `OPA_ptr_t` is defined. Moreover, I'd suggest define subroutine for such queue access operations similar to CH3 (see `MPID_nem_queue_enqueue` in CH3)

Alternatively, we could probably just make this `OPA_ptr_t` I think.

minsii commented on src/mpid/ch4/shm/posix/eager/iqueue/iqueue_recv.h:
```diff
@@ -0,0 +1,148 @@
+            }
+        }
+    }
+
```
I don't understand the logic of this function. Can you please explain why we need `first_cell` and `last_cell` at transport ? 

Besides, casting `OPA_ptr_t` type seems unsafe.

wesbland replied:
I believe they are used to tell the boundaries of a message that's larger than a single cell. At least, that's what the documentation I came up with says.

Is that true @SergeyGubanov?

wesbland replied:
> Besides, casting OPA_ptr_t type seems unsafe.

This will need the same solution as https://github.com/pmodels/mpich/pull/3199#discussion_r224264249

SergeyGubanov replied:
The terminal->head is a stack (inverted-ordered queue) of incomimg cells (public scope of visibility, visible to all ranks by shared memory).

The first_cell and last_cell is the true-ordered incoming queue (private scope of visibility, visible to the process which receive the cells).

minsii commented on src/mpid/ch4/shm/posix/eager/iqueue/iqueue_recv.h:
```diff
@@ -0,0 +1,148 @@
+/* Work backward through the queue to find the first cell and put that one in the transport's
+ * first_cell position. */
+static void MPIDI_POSIX_EAGER_IQUEUE_enqueue_cell(MPIDI_POSIX_EAGER_IQUEUE_transport_t * transport,
+                                                  MPIDI_POSIX_EAGER_IQUEUE_cell_t * cell)
```
Why the `MPIDI_POSIX_EAGER_IQUEUE_enqueue_cell` and `MPIDI_POSIX_EAGER_IQUEUE_receive_cell` functions use uppercase `EAGER_IQUEUE` ? Can we make them lowercase to be consistent with other functions ?

wesbland replied:
Done.

minsii commented on src/mpid/ch4/shm/posix/eager/iqueue/iqueue_recv.h:
```diff
@@ -0,0 +1,148 @@
+        } else {
+            MPIR_Assert(cell->type == MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_TAIL);
+            transaction->msg_hdr = NULL;
+        }
```
The naming of `MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_HEAD` and `MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_TAIL` seems confusing. It actually means whether it contains an AM header, but the type names read like location in a queue...Maybe we could change them to `...CELL_TYPE_HDR` and `...CELL_TYPE_DATA` ?

minsii replied:
`transaction` is not initialized.

wesbland replied:
> The naming of `MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_HEAD` and `MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_TAIL` seems confusing. It actually means whether it contains an AM header, but the type names read like location in a queue...Maybe we could change them to `...CELL_TYPE_HDR` and `...CELL_TYPE_DATA` ?

That sounds good to me. 👍

wesbland replied:
> `transaction` is not initialized.

Not sure what you mean here. The variable is created in [posix_progress.h](https://github.com/pmodels/mpich/pull/3199/files/8ff97b4aa32837c97d95c212e66dac64b45ae856..49167afc81da35f29869d1db4d8d0ba35cb41be3#diff-b7a55510bbf1a862e99ad3afd3df53e2R27) and passed in as a parameter. If a cell isn't found, `transaction` isn't used.

minsii replied:
>> transaction is not initialized.

>Not sure what you mean here. The variable is created in posix_progress.h and passed in as >a parameter. If a cell isn't found, transaction isn't used.

Yes, now I got it. Please just ignore this comment.

minsii commented on src/mpid/ch4/shm/posix/eager/iqueue/iqueue_send.h:
```diff
@@ -0,0 +1,136 @@
+            return cell;
+        }
+    }
+    return NULL;
```
Every time we have to search available cell from the head of the list, can we use the buffer pool structure to mange it (e.g., like request object) ?

SergeyGubanov replied:
It is a main feature. To free a cell only need to just set one flag. Synchronization doesn't needed. The total number of used cells always as small as possible (no extra cache pollution).

minsii replied:
I think a detailed documentation is needed to explain the iqueue design.

wesbland replied:
@SergeyGubanov - I think writing such detailed documentation goes beyond what I'm going to be able to produce given that I didn't write any of this code. Can you provide a document about the design (I can help format and/or figure out where to put it if you can come up with the content).

minsii commented on src/mpid/ch4/shm/posix/eager/iqueue/iqueue_send.h:
```diff
@@ -0,0 +1,136 @@
+                       MPIDI_POSIX_am_header_t ** msg_hdr,
+                       struct iovec **iov, size_t * iov_num, int is_blocking)
+{
+    MPIDI_POSIX_EAGER_IQUEUE_transport_t *transport;
```
Looks like we don't have `MPIR_FUNC_VERBOSE_STATE_DECL` set in all `fbox` and `iqueue` internal routines. Is it intentional ?

wesbland replied:
Probably not. I'll add them.

wesbland commented on src/mpid/ch4/src/ch4r_callbacks.h:
```diff
@@ -114,14 +118,61 @@ static inline int MPIDI_handle_unexp_cmpl(MPIR_Request * rreq)
+                }
+
+                break;
+            }
```
> I do not understand the logic of this while loop. Does it dequeue every matched-posted-request, and try to cancel the partner-request if it exists, then if the cancel fails, it dequeue and try-cancel the next matched-post-request ? Can you explain more ?

Again, I'm working on improving the documentation here:

This while loop is to ensure that if the first request dequeued from the posted list  turns out to be an MPI_ANY_SOURCE requests which was already matched by the other transport (netmod or shared memory), we can keep looking for the next request in the  posted queue.

wesbland replied:
> Is MPIR_Comm_release(root_comm); /* -1 for posted_list */ for the canceled anysource_partner ?

This is for this request's reference on the communicator. The partner will downref for itself when that request is finished.

minsii replied:
I am asking because I see two more downref at line 177. One is commented as for posted_list. Ah, is the above downref only performed for the request whose partner was not canceled successfully ? Maybe your documentation could help.

minsii commented on src/mpid/ch4/shm/posix/eager/iqueue/iqueue_recv.h:
```diff
@@ -0,0 +1,148 @@
+        MPIDI_POSIX_EAGER_IQUEUE_cell_t *prev =
+            MPIDI_POSIX_EAGER_IQUEUE_GET_CELL(transport, cell->prev);
+        MPIDI_POSIX_EAGER_IQUEUE_enqueue_cell(transport, prev);
+        cell->prev = 0;
```
Agreed that we can have both. Performance study is needed to verify how much overhead we pay for each queue.

wesbland commented on src/mpid/ch4/src/ch4r_callbacks.h:
```diff
@@ -302,8 +358,14 @@ static inline int MPIDI_recv_target_cmpl_cb(MPIR_Request * rreq)
+            MPIDI_CH4R_anysource_matched(MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(rreq),
+                                         MPIDI_CH4R_NETMOD, &continue_matching);
+        }
+
```
I think you're right here. I think this should check the result of `continue_matching` and bail out if necessary.

wesbland replied:
I take it back. This should have already happened in `MPIDI_send_target_msg_cb` so it's not necessary to check it here. Now I'm wondering why this code is necessary at all.

hzhou replied:
@wesbland I am looking at the code in current master and am wondering the same thing -- it appears that anysource_partner has already been resolved at matching time so this extra checking at `MPIDI_recv_target_cmpl_cb` is not necessary. Did you resolve your above comments back then?

wesbland replied:
Sorry, I don't remember the result of this. I don't have any internal code that's different than this though. So what you see in `master`, is what I have too.

wesbland commented on src/mpid/ch4/shm/include/shm.h:
```diff
@@ -424,6 +426,8 @@ typedef int (*MPIDI_SHM_mpi_iscatterv_t) (const void *sendbuf, const int *sendco
 
+/* These structs are used when inlining is turned off and we call functions pointers for the shared
+ * memory functions instead of directly inlining the functions into the device code. */
 typedef struct MPIDI_SHM_funcs {
```
👍 I found a couple more places where it made sense to pull these out.

wesbland commented on src/mpid/ch4/shm/posix/eager/fbox/fbox_init.h:
```diff
@@ -0,0 +1,164 @@
+               sizeof(MPIDI_POSIX_fastbox_t));
+
+        MPIDI_POSIX_eager_fbox_control_global.mailboxes.in[i]->flag = 0;
+    }
```
Done.

wesbland commented on src/mpid/ch4/shm/posix/eager/fbox/fbox_types.h:
```diff
@@ -0,0 +1,66 @@
+
+#define POSIX_FBOX_TRACE(...) \
+    MPL_DBG_MSG_FMT(MPIDI_CH4_SHM_POSIX_FBOX_GENERAL,VERBOSE,(MPL_DBG_FDEST, __VA_ARGS__))
+
```
I don't think that's true. If `MPL_USE_DBG_LOGGING` is off, the macros turn into nothing:

https://github.com/pmodels/mpich/blob/447f2fe7934446f387fec9704f494394b1d6d04f/src/mpl/include/mpl_dbg.h#L95-L105

minsii replied:
I meant the  `MPIDI_CH4_SHM_POSIX_FBOX_GENERAL` declaration.

wesbland replied:
`MPL_dbg_class` is also defined either way.

minsii replied:
But the `MPIDI_POSIX_eager_fbox_control_global` and `MPIDI_CH4_SHM_POSIX_FBOX_GENERAL` variables are always declared.

wesbland replied:
`MPIDI_POSIX_eager_fbox_control_global` isn't a debugging thing, but you're right about `MPIDI_CH4_SHM_POSIX_FBOX_GENERAL`. I'll put that in a check.

wesbland commented on src/mpid/ch4/shm/posix/posix_am.h:
```diff
@@ -3,37 +3,214 @@
-                                                  MPI_Count count,
-                                                  MPI_Datatype datatype, MPIR_Request * sreq)
+                                                  MPI_Count count, MPI_Datatype datatype,
+                                                  MPIR_Request * sreq)
```
Yeah, I wonder if the format script did something. I'll revert it.

wesbland commented on src/mpid/ch4/shm/posix/posix_am_impl.h:
```diff
@@ -0,0 +1,90 @@
+
+        req_hdr->am_hdr = MPL_malloc(am_hdr_sz, MPL_MEM_SHM);
+        MPIR_Assert(req_hdr->am_hdr);
+    }
```
Yep. There was another place in this file with the same thing.

wesbland commented on src/mpid/ch4/shm/posix/posix_am.h:
```diff
@@ -3,37 +3,214 @@
+        } else {
+            curr_sreq_hdr->iov[0] = iov_left_ptr[0];
+        }
+    }
```
I'm not 100% sure, but I think the problem here could be that you are sending an AM header-only message (`data_sz == 0`) and `curr_sreq_hdr` is the pointer to the AM header while `iov_left_ptr` is for the actual data. If you have some header left to send, you need to make sure you're sending the right thing.

wesbland commented on src/mpid/ch4/shm/posix/eager/fbox/fbox_recv.h:
```diff
@@ -0,0 +1,149 @@
+                transaction->payload_sz = fbox_in->payload_sz - sizeof(MPIDI_POSIX_am_header_t);
+            } else {
+                /* Received a fragment fragment of the message payload */
+                transaction->msg_hdr = NULL;
```
Done.

wesbland commented on src/mpid/ch4/shm/posix/eager/fbox/fbox_recv.h:
```diff
@@ -0,0 +1,149 @@
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_POSIX_FBOX_EAGER_RECV_BEGIN);
+
+    enum { BATCH_SIZE = 4 };
+    int j, local_rank, grank;
```
Yeah, I think that's a good idea.

wesbland commented on src/mpid/ch4/shm/posix/eager/fbox/fbox_recv.h:
```diff
@@ -0,0 +1,149 @@
+    int j, local_rank, grank;
+    MPIDI_POSIX_fastbox_t *fbox_in;
+    int mpi_errno = MPIDI_POSIX_NOK;
+
```
You mean we should set the values to `NULL` (or whatever makes sense) when we don't find something? I'm not sure that's necessary since we indicate that with the return value.

minsii replied:
Oh, I missed the return value part. Then it is OK.

wesbland commented on src/mpid/ch4/shm/posix/eager/fbox/fbox_recv.h:
```diff
@@ -0,0 +1,149 @@
+    MPIR_Memcpy(dst, src, size);
+
+    return;
+}
```
Done.

wesbland commented on src/mpid/ch4/shm/posix/eager/fbox/fbox_recv.h:
```diff
@@ -0,0 +1,149 @@
+
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_POSIX_FBOX_EAGER_RECV_COMMIT);
+    return;
+}
```
Done.

wesbland commented on src/mpid/ch4/shm/posix/eager/iqueue/iqueue_types.h:
```diff
@@ -0,0 +1,97 @@
+    MPIDI_POSIX_EAGER_IQUEUE_cell_t *last_cell; /* Internal variable to track the last cell to be
+                                                 * received when receiving multiple cells. */
+} MPIDI_POSIX_EAGER_IQUEUE_transport_t;
+
```
Done.

wesbland commented on src/mpid/ch4/shm/posix/eager/iqueue/iqueue_recv.h:
```diff
@@ -0,0 +1,148 @@
+                              void *dst, const void *src, size_t size)
+{
+    memcpy(dst, src, size);
+}
```
Done.

wesbland commented on src/mpid/ch4/shm/posix/eager/iqueue/iqueue_send.h:
```diff
@@ -0,0 +1,136 @@
+            /* Optimize for the case where the message will fit into the cell. */
+            if (likely(available >= (*iov)[i].iov_len)) {
+                memcpy(payload, (*iov)[i].iov_base, (*iov)[i].iov_len);
+
```
Done.

wesbland commented:
Ok. I finished the first round of responses to the review. Thanks for all the feedback @minsii. I think there's a few things left that need to be resolved either through discussion between us or with @SergeyGubanov.

wesbland commented:
I've also force pushed my changes to this branch so many of the comments above will probably now be marked as "outdated".

hajimefu commented on src/mpid/ch4/src/ch4_recv.h:
```diff
@@ -136,19 +139,32 @@ MPL_STATIC_INLINE_PREFIX int MPID_Recv_init(void *buf,
-        if (mpi_errno != MPI_SUCCESS) {
-            MPIR_ERR_POP(mpi_errno);
-        }
+        if (!MPIR_Request_is_complete(*request) && !MPIDI_CH4U_REQUEST_IN_PROGRESS(*request)) {
```
I have two concerns in this line:
1. This might not be quite correct from the patch ordering point of view -- MPIDI_CH4U_REQUEST_IN_PROGRESS assumes that the shmmod uses the active message code but it does not at this moment.
2. Also, is checking for MPIR_Request_is_complete(*request) valid here? Here *request is an inactive persistent request, which is always considered as "completed" if I understand correctly. So, this condition will completely disable NM path.

wesbland replied:
Oops. I copied your comments over as well. 😄 See my responses below.

wesbland commented on src/mpid/ch4/src/ch4_recv.h:
```diff
@@ -136,19 +139,32 @@ MPL_STATIC_INLINE_PREFIX int MPID_Recv_init(void *buf,
-        if (mpi_errno != MPI_SUCCESS) {
-            MPIR_ERR_POP(mpi_errno);
-        }
+        if (!MPIR_Request_is_complete(*request) && !MPIDI_CH4U_REQUEST_IN_PROGRESS(*request)) {
```
From #3339 by @hajimefu:

> Also, is checking for `MPIR_Request_is_complete(*request)` valid here? Here `*request` is an inactive persistent request, which is always considered as "completed" if I understand correctly.

I think this might be a valid point. Perhaps this should drop the completion check and just look at whether or not it's in progress.

wesbland replied:
Removed the completion check.

SergeyGubanov replied:
MPIDI_CH4U_REQUEST_IN_PROGRESS caused segmentation fault when request is completed (due to null pointer dereference inside MPIDI_CH4U_REQUEST_IN_PROGRESS).

To avoid segmentation fault, in the first we must check MPIR_Request_is_complete() and if that is not true, we can check MPIDI_CH4U_REQUEST_IN_PROGRESS().

hajimefu commented on src/mpid/ch4/src/ch4_recv.h:
```diff
@@ -136,19 +139,32 @@ MPL_STATIC_INLINE_PREFIX int MPID_Recv_init(void *buf,
+            MPIDI_CH4I_REQUEST(MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(*request), is_local) = 0;
+
+            /* cancel the shm request if netmod/am handles the request from unexpected queue. */
+            if (MPIR_Request_is_complete(MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(*request))) {
```
Same argument here: I don't think it makes sense to check for request completion here. (I would make sense for regular receive, but remember this line is for persistent request.)

wesbland replied:
I think this one is easier since the request will never be complete, we can just remove this code.

hajimefu replied:
Yes, that's what I would do.

SergeyGubanov replied:
It is not true. 
This line for the netmod request MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(*request).
Netmod request can be completed after the call MPIDI_NM_mpi_irecv(buf, count, datatype, rank, tag, comm, context_offset, av, &(MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(*request)));

wesbland replied:
@SergeyGubanov - I think we're talking about the persistent call, `MPID_Recv_init`, rather than the regular receive, `MPID_Recv`.

`MPIDI_NM_mpi_irecv` can definitely complete the request, but I don't know that `MPIDI_NM_mpi_recv_init` is allowed to carry out the full receive operation, particularly since the user hasn't yet called `MPI_Start` on it.

As @hajimefu mentioned above, the request cannot be completed at this point (you only call `MPI_Recv_init` once so having the request be freed shouldn't be a problem). In fact, I'm wondering if the call to `MPIDI_CH4U_REQUEST_IN_PROGRESS` is necessary as the request can't be in progress at this point either. This is really just to set up the request object.

hajimefu replied:
@wesbland @SergeyGubanov Yeah, so to summarize, this kind of logic is necessary but it should be in `MPID_Startall`, not here (`MPID_Recv_init`). I happen to be working with `ANY_SOURCE` receive in persistent communication and now fixing it in `MPID_Startall`. I'll upstream the patch soon.

wesbland commented on src/mpid/ch4/src/ch4_recv.h:
```diff
@@ -136,19 +139,32 @@ MPL_STATIC_INLINE_PREFIX int MPID_Recv_init(void *buf,
-        if (mpi_errno != MPI_SUCCESS) {
-            MPIR_ERR_POP(mpi_errno);
-        }
+        if (!MPIR_Request_is_complete(*request) && !MPIDI_CH4U_REQUEST_IN_PROGRESS(*request)) {
```
From #3339 by @hajimefu:

> This might not be quite correct from the patch ordering point of view -- `MPIDI_CH4U_REQUEST_IN_PROGRESS` assumes that the shmmod uses the active message code but it does not at this moment.

I'm not sure this is true. It assumes that the _request_ is using, but I don't think this patch has any shared memory requirements. It won't do much at this point because shared memory don't yet use active messages. But it if did, this code would apply.

hajimefu replied:
Let me rephrase my point. `MPIDI_CH4U_REQUEST_IN_PROGRESS` checks `(r)->dev.ch4.am.req->status`. What if shmmod has no idea about the active message (i.e. `(r)->dev.ch4.am.req`), which is true in the current `master`.

Note that `(r)->dev.ch4.am` and `(r)->dev.ch4.shm` are under the same union and you cannot use both at the same time.

wesbland replied:
> Note that (r)->dev.ch4.am and (r)->dev.ch4.shm are under the same union and you cannot use both at the same time.

Hmm. This is a good point. I'll need to look at that.

hajimefu replied:
After merging the entire PR this might not be a huge deal, but I noticed this because I was cherry picking this particular commit for my experiment.

wesbland replied:
Yeah, at this point I'm convinced that the ordering here should change, but I think the union thing you mentioned above is actually a problem that needs to be resolved.

wesbland replied:
I've reordered the patches to avoid the patch ordering problem and moved the fields in the `MPIDI_POSIX_request_t` struct to `MPIDI_POSIX_am_request_t` so they won't have the union issues.

wesbland commented:
I did a quick rebase and fixed a problem with header inclusion ordering with non-inlined builds. Things should be in better shape now. I'll look at the `OPA_ptr_t` issue above and then we should be ready to go code-wise I think. Hopefully @SergeyGubanov can get back to us on the comment above.

wesbland commented:
test-intel:ch4/ofi/sockets/gnu,icc/force-am-disabled/auto,no-odd-even/debug,default

hajimefu commented:
@wesbland `./f90/pt2pt/mprobef90 2` is failing due to the rebase. I have a patch.

wesbland commented:
test:intel:ch4/ofi/sockets/gnu,icc/force-am-disabled/auto,no-odd-even/debug,default

wesbland commented:
After a discussion with @minsii at SC, we think the best course of action here is to drop the iqueue eager module for now. The performance isn't significantly better than the fastbox code that's also part of this PR and the added complexity is still not well understood.

The plan is to port over more of the optimizations present in the CH3 fastbox implementation in the near future and continue working on further optimizations at the CH4 level, which should make the fastbox implementation better than the iqueue implementation anyway.

For now at least, the iqueue code will still be maintained internally by Intel so that we can upstream the code again later if we think it's helpful.

hajimefu commented:
@halimamer You may want to look at these two patches I added here.
* ch4: Initialize SHM AM request as well
* ch4: Remove ref_increment parameter from am_request_init 

They fix (a part of) crashes that show up in trylock after merging this shared memory PR.

wesbland commented:
test:jenkins/ch4/most

wesbland commented:
test:intel:ch4/ofi/sockets/gnu,icc/force-am-disabled,force-am-enabled/auto,no-odd-even,thread/debug,default

wesbland commented:
test:intel:ch4/ofi/sockets/gnu,icc/force-am-disabled,force-am-enabled/auto,no-odd-even,thread/debug,default

wesbland commented:
test:intel:ch4/ofi/sockets/gnu,icc/force-am-disabled,force-am-enabled/auto,no-odd-even,thread/debug,default

wesbland commented:
The remaining failures in the ch4/ofi job and the intel testing jobs are resolved by other PRs, particularly #3383.

This PR no longer has the `iqueue` code and `fbox` is again the default.

minsii commented:
@wesbland : we should update the commit message in `ch4/shm: POSIX shared memory new logic`, i.e., remove iqueue

halimamer commented:
@hajimefu , got it, will take a look

minsii commented:
@wesbland I believe the active message hanging issue with `fbox` was resolved by adding the queueing logic in `MPIDI_POSIX_am_send_hdr`.

halimamer commented:
@hajimefu , the changes look fine to me

wesbland commented:
Fixed the rebase conflicts.

test:jenkins/ch4/most

raffenet commented:
OK, I think this is good to go. Last step is to add signoffs. I can do it, just as long as someone can let me know who I should be adding.
